### ISSUE TITLE
Question about non-causal mask

### ISSUE THREAD

**Issue by mmorinag127** [nebius/kvax#2]
Dear authors, 

This is incredible! Thank you for your excellent work!
I have a straightforward question about the masking without a causal flag.
Can it be utilized for both causal and non-causal attention?
If it can, could you please tell me how to do it?

Thanks a lot!

**Comment by southfreebird** [comment]
Hi!

Thank you for your feedback!

The causal or positional mask is built based on the `query_positions` and `kv_positions` tensors using a simple condition: `query_positions >= kv_positions`.

So, setting:
```
query_positions = jnp.broadcast_to(jnp.arange(query_seq_len), (batch_size, query_seq_len))
kv_positions = jnp.broadcast_to(jnp.arange(kv_seq_len), (batch_size, kv_seq_len))
```
results in the creation of a causal mask, whereas
```
query_positionals = jnp.zeros((batch_size, query_seq_len), dtype=jnp.int32)
kv_positionals = jnp.zeros((batch_size, kv_seq_len), dtype=jnp.int32)
```
leads to the calculation of full tokens (without a causal mask). This approach can also be used to create custom positional masks.

For more details, please refer to the "Computation of Document Mask" section in the blog post: [Kvax blog post](https://nebius.com/blog/posts/kvax-open-source-flash-attention-for-jax#computation-of-document-mask)

I think this is a great point to highlight in the documentation.
Please feel free to ask if anything is unclear.

**Comment by mmorinag127** [comment]
Hi!

Thanks a lot! Now I understand what it is.
Then I have a question about the dtype of q/k/v at the backward.
[here](https://github.com/nebius/kvax/blob/main/kvax/ops/flash_attention_triton.py#L1728)

```python
    dquery_shape_dtype = jax.ShapeDtypeStruct(shape=query.shape, dtype=query.dtype)
    dkey_shape_dtype = jax.ShapeDtypeStruct(shape=dk_shape, dtype=jnp.float32)
    dvalue_shape_dtype = jax.ShapeDtypeStruct(shape=dv_shape, dtype=jnp.float32)
```
Why do you set different types for query and kv?
I'm wondering if kv.dtype can be promoted to query.dtype at the end of the function

Thanks!

**Comment by southfreebird** [comment]
Hi @mmorinag127,

This was done on purpose to perform summarisation by heads in the case of GQA with high precision ([here](https://github.com/nebius/kvax/blob/main/kvax/ops/flash_attention_triton.py#L1857-L1868)), and more importantly, to perform the resulting `AllReduce` in high precision when context parallelism is enabled.
If you change it to lower precision at the end of the function, it causes `AllReduce` to run in lower precision, which leads to an accuracy drop.

But we did notice it causing some issues, so we’ve prepared a fix. I’ll merge it into our code shortly.

**Comment by southfreebird** [comment]
I have added a new commit to main that converts `dkey` and `dvalue` to the `query`'s dtype without affecting the precision of `AllReduce`. For more details, take a look at [this](https://github.com/nebius/kvax/commit/d4ef9ece8fc4c3373e3d9942f8be2dfad1603696) commit.
